pkg/telnetlib/connection.go

```diff
@@ -0,0 +1,316 @@
+// Copyright 2016 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package telnetlib
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"net"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+type connOpts struct {
+	conn        net.Conn
+	fsm         *telnetFSM
+	cmdHandler  CmdHandlerFunc
+	dataHandler DataHandlerFunc
+	serverOpts  map[byte]bool
+	clientOpts  map[byte]bool
+	optCallback func(byte, byte)
+}
+
+type TelnetConn struct {
+	conn              net.Conn
+	readCh            chan []byte
+	writeCh           chan []byte
+	unackedServerOpts map[byte]bool
+	unackedClientOpts map[byte]bool
+	//server            *TelnetServer
+	serverOpts         map[byte]bool
+	clientOpts         map[byte]bool
+	dataRW             io.ReadWriter
+	cmdBuffer          bytes.Buffer
+	fsm                *telnetFSM
+	fsmInputCh         chan byte
+	handlerWriter      io.Writer
+	cmdHandler         CmdHandlerFunc
+	dataHandler        DataHandlerFunc
+	dataHandlerCloseCh chan chan struct{}
+	dataWrittenCh      chan bool
+	optionCallback     func(byte, byte)
+	connReadDoneCh     chan chan struct{}
+	connWriteDoneCh    chan chan struct{}
+	negotiationDone    chan struct{}
+	closedMutex        sync.Mutex
+	closed             bool
+}
+
+// Safely read/write concurrently to the data Buffer
+// databuffer is written to by the FSM and it is read from by the dataHandler
+type dataReadWriter struct {
+	dataBuffer bytes.Buffer
+	dataMux    *sync.Mutex
+}
+
+func (drw *dataReadWriter) Read(p []byte) (int, error) {
+	drw.dataMux.Lock()
```

Comment:
I'm guessing we can't read from the dataReadWriter AND write to it at the same time?  Can we use a RW lock and update a cursor so we can call Read more than once?